### PR TITLE
[IMP] base: validate mail server SSL certificates (stable)

### PR DESCRIPTION
**Pull request by Julien00859**
The default SSL context created by Python in the `SMTP_SSL` class and
the `starttls` method is very permissive: it doesn't validate the
server certificate, nor its hostname. Python would accept any
self-signed certificate even for well known services such as Google or
Microsoft making man-in-the-middle attacks possible.

This is not *as bad* as it sounds: the connections are established by
the server, not the clients. Servers usually sit in a datacenter and
have enhanced security (static IP, static DNS, firewall, ...) making
mitm attacks unlikely. Still, it is a very bad practice to just accept
any certificate: the connection is only safe when both encryption and
authentification are in place.

This work adds two new encryption methods, namely "ssl_strict"
and "starttls_strict". Both act exactly the same as their "not strict"
variant with the exception that they validate peers certificates. They
validate that (1) the peer certificate is signed by a trusted authority
and that (2) the Subject Alternative Name (SAN) of the certificate
matches the smtp server hostname.

It was not possible to "just" enforce those validation on the
existing "ssl" and "starttls" encryption methods: there are many smtp
servers in the wild that use expired, or self-signed, or
hostname-mismatching certificates. The connection to those server is
still encrypted if not authenticated. We considere it is much better
than clear-text and breaking the connection to those servers would be
too disruptive to the customers.

task-2861790

### CODE REVIEW

File: odoo/tools/config.py [hunk 1]
```diff
@@ -650,6 +658,14 @@ def _test_enable_callback(self, option, opt, value, parser):
         if not parser.values.test_tags:
             parser.values.test_tags = "+standard"
 
+    def _smtp_ssl(self, option, opt, value, parser):
+        encryption_methods = ('none', 'ssl', 'ssl_strict', 'starttls', 'starttls_strict')
+        if parser.rargs and parser.rargs[0] in encryption_methods:
+            parser.values.smtp_ssl = parser.rargs.pop(0)
+        else:
+            # backward compatibility, option used to be a flag
+            parser.values.smtp_ssl = 'starttls'
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/config.py:667]: Yup I did test it.
```
$ o --log-level=warn --stop
/home/julien/.pyenv/versions/odoo17/bin/python3 community/odoo-bin --addons-path=~/Odoo/custo,~/Odoo/design-themes,~/Odoo/enterprise,~/Odoo/community/odoo/addons,~/Odoo/community/addons server --log-level=warn --stop-after-init -c config.cfg -d odoo-base-ir_mail_server-strict-stable --db-filter ^odoo-base-ir_mail_server-strict-stable$
self.options['smtp_ssl']='none'
```
```
$ o --log-level=warn --stop --smtp-ssl
/home/julien/.pyenv/versions/odoo17/bin/python3 community/odoo-bin --addons-path=~/Odoo/custo,~/Odoo/design-themes,~/Odoo/enterprise,~/Odoo/community/odoo/addons,~/Odoo/community/addons server --log-level=warn --stop-after-init --smtp-ssl -c config.cfg -d odoo-base-ir_mail_server-strict-stable --db-filter ^odoo-base-ir_mail_server-strict-stable$
self.options['smtp_ssl']='starttls'
```
```
$ o --log-level=warn --stop --smtp-ssl ssl_strict
/home/julien/.pyenv/versions/odoo17/bin/python3 community/odoo-bin --addons-path=~/Odoo/custo,~/Odoo/design-themes,~/Odoo/enterprise,~/Odoo/community/odoo/addons,~/Odoo/community/addons server --log-level=warn --stop-after-init --smtp-ssl ssl_strict -c config.cfg -d odoo-base-ir_mail_server-strict-stable --db-filter ^odoo-base-ir_mail_server-strict-stable$
self.options['smtp_ssl']='ssl_strict'
```
```
$ o --log-level=warn --stop --smtp-ssl toto
/home/julien/.pyenv/versions/odoo17/bin/python3 community/odoo-bin --addons-path=~/Odoo/custo,~/Odoo/design-themes,~/Odoo/enterprise,~/Odoo/community/odoo/addons,~/Odoo/community/addons server --log-level=warn --stop-after-init --smtp-ssl toto -c config.cfg -d odoo-base-ir_mail_server-strict-stable --db-filter ^odoo-base-ir_mail_server-strict-stable$
Usage: odoo-bin server [options]

odoo-bin server: error: unrecognized parameters: 'toto'
```

[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/config.py:667]: I also tested  adding `nargs=0` on the `add_option`. It works too. Nothing else was changed. Do you want I add it?

File: odoo/addons/base/models/ir_mail_server.py [hunk 2]
```diff
@@ -337,11 +373,29 @@ def connect(self, host=None, port=None, user=None, password=None, encryption=Non
...
+                ssl_context.check_hostname = True
+                ssl_context.verify_mode = ssl.CERT_REQUIRED
+            else:
+                ssl_context = ssl.create_default_context()
+                ssl_context.check_hostname = False
+                ssl_context.verify_mode = ssl.CERT_NONE
             if (mail_server.smtp_authentication == "certificate"
                and mail_server.smtp_ssl_certificate
                and mail_server.smtp_ssl_private_key):
                 try:
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/addons/base/models/ir_mail_server.py:389]: At the moment the `smtp_authentication="login"` is used for both login auth (when smtp_user is set) and no auth (when smtp_user is null). We would need a `smtp_authentication="none"` for an exception to be useful.

File: odoo/addons/base/models/ir_mail_server.py [hunk 3]
```diff
@@ -83,6 +92,24 @@ def extract_rfc2822_addresses(text):
...
+       Subject Alternative Name (SAN)."""
+    if err_no:
+        return False
+
+    if err_depth == 0:  # leaf certificate
+        peercert = {
+            "subject": ((("commonName", x509.get_subject().CN),),),
+            "subjectAltName": get_subj_alt_name(x509),
+        }
+        match_hostname(peercert, hostname)  # it raises when it does not match
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/addons/base/models/ir_mail_server.py:108]: Raising inside the function allow us to catch the exception rose by `match_hostname` and to show a valuable error message to the user:
```
An SSL exception occurred. Check connection security type.
 CertificateError: hostname 'notlocalhost' doesn't match 'localhost'
```
If we are to `try: match_hostname(...); except CertificateError: return False` we loose the information and a cryptic message is shown:

```
An SSL exception occurred. Check connection security type.
 ("bad handshake: Error([(\'SSL routines\', \'tls_process_server_certificate\', \'certificate verify failed\')])",)
```

That pyOpenSSL handles the situation is very nice and the day it doesn't anymore the test will blow up and we'll know something is fishy.

**Comment by Julien00859** [comment]
163584

File: odoo/addons/base/models/ir_mail_server.py [hunk 4]
```diff
@@ -337,11 +373,29 @@ def connect(self, host=None, port=None, user=None, password=None, encryption=Non
             smtp_encryption = mail_server.smtp_encryption
             smtp_debug = smtp_debug or mail_server.smtp_debug
             from_filter = mail_server.from_filter
+            if mail_server.smtp_authentication == 'none':
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/addons/base/models/ir_mail_server.py]: was a bug, fixed

File: odoo/addons/base/models/ir_mail_server.py [hunk 5]
```diff
@@ -83,6 +92,24 @@ def extract_rfc2822_addresses(text):
...
+       :param:`hostname` matches the :param:`x509`'s Common Name (CN) or
+       Subject Alternative Name (SAN)."""
+    if err_no:
+        return False
+
+    if err_depth == 0:  # leaf certificate
+        peercert = {
+            "subject": ((("commonName", x509.get_subject().CN),),),
+            "subjectAltName": get_subj_alt_name(x509),
+        }
```
Comment (std-odoo) [inline odoo/addons/base/models/ir_mail_server.py:107]: can't we get this from the socket instead ? (with `getpeercert`)

**Review by std-odoo — COMMENTED** [review]
@Julien00859 Hi :)

Thanks for making the SSL feature not useless :)

Globally look good for me, I have a few picky comments, and 1 or 2 questions

Cheers

File: odoo/addons/base/models/ir_mail_server.py [hunk 6]
```diff
@@ -17,13 +18,21 @@
...
-from urllib3.contrib.pyopenssl import PyOpenSSLContext
+from OpenSSL.SSL import Error as SSLError, VERIFY_PEER, VERIFY_FAIL_IF_NO_PEER_CERT
+from urllib3.contrib.pyopenssl import PyOpenSSLContext, get_subj_alt_name
 
 from odoo import api, fields, models, tools, _
 from odoo.exceptions import UserError
 from odoo.tools import ustr, pycompat, formataddr, email_normalize, encapsulate_email, email_domain_extract, email_domain_normalize
 
+try:
+    from urllib3.util.ssl_match_hostname import CertificateError, match_hostname
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/addons/base/models/ir_mail_server.py:30]: urllib3 versions we care about are:
* 1.24 (debian buster, python3.7, odoo15)
* 1.25 (ubuntu focal)
* 1.26 (debian bullseye/bookworm, ubuntu jammy)
* 2.0 (ubuntu noble)

The function originated in Python 3.3 as `ssl.match_hostname` but has been deprecated in Python 3.7. It supports only domain names and not ip addresses.

For ages the urllib3 team did vendor the function, I have been able to retrace its inclusion to 2011. The function was vendored as-is until ~2016. Inside a version older than 1.24 the urllib team started to take ownership on that function and added a bunch of fixes/patches, e.g. to support IP Address inside Subject Alternative Name (x509 extension). So we wanna use *their* `match_hostname` for the added features.

In 1.24/1.25 the function should be imported from `urllib3.packages.ssl_match_hostname`. They moved it to `urllib3.util.ssl_match_hostname` in 1.26.

That last fallback on the python's function is actually useless because urllib already is smart about it:

https://github.com/urllib3/urllib3/blob/9b95f29c575d73260ec4f5a8c0ea368cf242019e/src/urllib3/packages/ssl_match_hostname/__init__.py#L3-L16

File: odoo/addons/base/models/ir_mail_server.py (hunk 5 above)
Comment (Julien00859) [reply odoo/addons/base/models/ir_mail_server.py:106]: I guess you are talking about [getpeercert](https://docs.python.org/3/library/ssl.html#ssl.SSLSocket.getpeercert) from the stdlib. Here we are using pyopenssl and it seems that there is no such function https://www.pyopenssl.org/en/latest/search.html?q=getpeercert

File: odoo/addons/base/models/ir_mail_server.py [hunk 7]
```diff
@@ -153,14 +153,14 @@ def _compute_smtp_authentication_info(self):
...
-    @api.constrains('smtp_ssl_certificate', 'smtp_ssl_private_key')
+    @api.constrains('smtp_authentication', 'smtp_ssl_certificate', 'smtp_ssl_private_key')
     def _check_smtp_ssl_files(self):
-        """We must provided both files or none."""
         for mail_server in self:
-            if mail_server.smtp_ssl_certificate and not mail_server.smtp_ssl_private_key:
-                raise UserError(_('SSL private key is missing for %s.', mail_server.name))
-            elif mail_server.smtp_ssl_private_key and not mail_server.smtp_ssl_certificate:
-                raise UserError(_('SSL certificate is missing for %s.', mail_server.name))
+            if mail_server.smtp_authentication == 'certificate':
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/addons/base/models/ir_mail_server.py:191]: I did think about it but I don't think it is a good idea.

Let's take a noob user who doesn't understand shit about oauth/certificates/tls. He is struggling on the ir.mail_server form view. He is changing the `smtp_authentication` back and forth between `login` and `certificate` trying to understand what is going on.

At the moment he can leave the two `smtp_ssl_private_key` and `smtp_ssl_certificate` dangling. In case he tries with `smtp_authentication='login'` those two fields are simply ignored. If he goes back to `smtp_authentication='certificate'` the two fields just continue to hold their previous value, he doesn't have to supply them again.

With the change you suggest, an UserError would show up asking him to remove those two files... but since he is on `smtp_authentication != 'certificate'` neither field would show up on the form view. He would need first to select `smtp_authentication='certificate'` to be able to remove the two fields and then go back on `login` to test his shit. In case he changes is mind again and come back on `smtp_authentication='certificate'` then he would have to supply both fields again...

I don't think it is a good idea :D

File: odoo/addons/base/models/ir_mail_server.py [hunk 8]
```diff
@@ -337,11 +373,20 @@ def connect(self, host=None, port=None, user=None, password=None, encryption=Non
             smtp_encryption = mail_server.smtp_encryption
             smtp_debug = smtp_debug or mail_server.smtp_debug
             from_filter = mail_server.from_filter
-            if (mail_server.smtp_authentication == "certificate"
-               and mail_server.smtp_ssl_certificate
-               and mail_server.smtp_ssl_private_key):
+
+            if mail_server.smtp_encryption == 'none':
+                pass
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/addons/base/models/ir_mail_server.py:377]: but... I like it this way :'(

File: odoo/addons/base/models/ir_mail_server.py [hunk 9]
```diff
@@ -337,11 +373,20 @@ def connect(self, host=None, port=None, user=None, password=None, encryption=Non
...
+            elif mail_server.smtp_authentication == "certificate":
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/addons/base/models/ir_mail_server.py:378]: It should be fully covered by the unittests! I also did test various things manually (but using the same software I used for the tests, I confess) using the following script that I wrote for myself: https://gist.github.com/Julien00859/5090d1cff6c02197e5854aabb67bf5ac .

File: odoo/addons/base/models/ir_mail_server.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/addons/base/models/ir_mail_server.py:30]: I did add the missing comments

File: odoo/addons/base/models/ir_mail_server.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (std-odoo) [reply odoo/addons/base/models/ir_mail_server.py:191]: or why not keeping the old condition then, when `mail_server.smtp_authentication != 'certificate'` ? (both should be set or none)

File: odoo/addons/base/models/ir_mail_server.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/addons/base/models/ir_mail_server.py]: changed

**Comment by Julien00859** [comment]
@robodoo rebase-ff

**Comment by Julien00859** [comment]
Hello @odoo/rd-security @tde-banana-odoo :wave:

TL;DR: I'm noticing you that I'm going to self-r+ this PR, it became urgent with py3.13.

The default `ssl_context` created by python on-the-fly when using the `SMTP` and `SMTP_SSL` classes is unsafe by default. It does not verify nor the certificate nor the hostname of the remote smtp server. This changed just now in 3.13, Python now enforces the verification.

It means that we have customers who are able to use their (broken) postfix server with Odoo for years... But when they upgrade to py3.13 (or we do on odoo online/sh), their emails won't send anymore. This will be very disruptive for them (and for the support).

This PR does 2 things: (1) It doesn't let the `SMTP` and `SMTP_SSL` come up with a default `ssl_context`, we now always provide one (so the behavior doesn't change with py3.13). (2) It offers two new selection values `ssl_strict` and `starttls_strict` to enforce the certificate/hostname verification.

Everything has been carefully implemented and there is an entire test suite covering 100% of the cases. @std-odoo and @xmo-odoo had a look already. I'm going to merge this afternoon unless you speak out.

**Comment by Julien00859** [comment]
> That seems excessive. We don't even know if our dependencies run on 3.13 let alone safe_eval.

Noted. Indeed I can't manage to have a 3.13 running with Odoo.

> To be clear: as in the default context sets check_certificate (or whatever) by default as of 3.13?

Indeed that what I meant.

**Comment by Julien00859** [comment]
@robodoo r+
